deeppavlov/core/models/tf_backend.py

```diff
@@ -20,28 +20,58 @@
 
 
 def _graph_wrap(func, graph):
+    """Constructs function encapsulated in the graph."""
     @wraps(func)
     def _wrapped(*args, **kwargs):
         with graph.as_default():
             return func(*args, **kwargs)
     return _wrapped
 
 
+def _keras_wrap(func, graph, session):
+    """Constructs function encapsulated in the graph and the session."""
+    import keras.backend as K
+
+    @wraps(func)
+    def _wrapped(*args, **kwargs):
+        with graph.as_default():
+            K.set_session(session)
+            return func(*args, **kwargs)
+    return _wrapped
+
+
 class TfModelMeta(with_metaclass(type, ABCMeta)):
-    """Metaclass that helps all child classes to have their own graph."""
+    """Metaclass that helps all child classes to have their own graph and session."""
     def __call__(cls, *args, **kwargs):
+        obj = cls.__new__(cls)
+
         from .keras_model import KerasModel
         if issubclass(cls, KerasModel):
             import keras.backend as K
+            if K.backend() != 'tensorflow':
+                obj.__init__(*args, **kwargs)
+                return obj
+
             K.clear_session()
+            obj.graph = tf.Graph()
+            with obj.graph.as_default():
+                if hasattr(cls, '_config_session'):
+                    obj.sess = cls._config_session()
+                else:
+                    obj.sess = tf.Session()
+                K.set_session(obj.sess)
```

Comment:
What is the reason to put another `K.set_session` here?
We already have it in every method.

Reply:
No reason really.